Statemanager: Cache Rewrite (Account/Storage) & Refactoring (Interface/Tests) (old)

holgerd77 opened the pull request:
This PR replaces the previous two cache rewrite PRs #2528 - allowing for cache deactivation - and #2590 - introducing a broad-scale refactor with a separation of access and diff (account) cache and preparing for a subsequent deeper cache integration and persistance.

PR is now independent from #2581 and targeting `master`, so that these two improvements can be handled and reviewed independently. Changes from here still fix the recursive DELEGATECALL out-of-memory issue from block 2283516.

Currently there is still a bug in the code since some blocks are quitting with a receipt root error. I have added a more extensive dedicated test suite for checkpointing tests to StateManager, this helped but something is still going wrong. Eventually the CI run here will point to something, since interdependencies with the other PR are now removed.

Additionally PR contains two small fixes:

- Bug in `StateManager.copy()` not taking in the options
- ~Bug in `Util` `Account.isEmtpy()` ignoring empty storage root~ 

jochem-brouwer commented on packages/util/src/account.ts:
```diff
@@ -122,7 +122,12 @@ export class Account {
+      this.balance === _0n &&
+      this.nonce === _0n &&
+      this.codeHash.equals(KECCAK256_NULL) &&
+      this.storageRoot.equals(KECCAK256_RLP)
```
I remember changing this in the past as well too. This breaks tests. I think it is related to gas costs, if all fields are zero except `storageRoot` then the account, for EVM purposes, is still empty. 

If you remove this line, `npm run test:state -- --fork=London` changes from `fail  1778` to `fail  186`

holgerd77 replied:
Have reverted.

jochem-brouwer commented on packages/statemanager/src/cache.ts:
```diff
@@ -14,74 +15,176 @@ export interface CacheOpts {
+    if (account.isEmpty()) {
+      return false
+    } else {
+      return true
```
I am confused here, if the account is not empty, the key is deleted?

jochem-brouwer commented on packages/vm/src/eei/vmState.ts:
```diff
@@ -266,6 +266,7 @@ export class VmState implements EVMStateAccess {
       }
     }
     await this._stateManager.flush()
+    this.touchedJournal.clear()
```
I think we need to investigate this a bit further. The problem with the client test failing (`getTransactionReceipt.spec.ts`) was that:

- The blockchain is initialized with a Geth genesis state
- This creates accounts for a big precompile range (0x00 - 0x00..00ff)`
- However, when creating genesis (in `vmexecution.open()`), these accounts are all created and dumped into `touchedJournal`. This is /not/ cleared
- However, on current master, the touched accounts are empty

This sounds a bit error prone to me and I want to investigate why this change has happened.

holgerd77 commented:
Rebased this and removed the `mainnetOptimizations` option (didn't work on follow-up blocks due to SELFDESTRUCT being executed repeatedly with a beneficiary in the targeted zero-based address range).

jochem-brouwer commented on packages/client/lib/execution/vmexecution.ts:
```diff
@@ -48,6 +54,10 @@ export class VMExecution extends Execution {
         trie,
+        cacheOptions: {
+          deactivate: false,
+          size: 2500000,
```
This immediately allocates these items, do we want this?

holgerd77 replied:
Hmm, yeah, maybe this can get optimized, I would also say that it doesn't hurt too much right now (and I am also not so motivated to improve here atm).

Are you generally sure about this behavior? Where did you "see" that?

jochem-brouwer replied:
![image](https://user-images.githubusercontent.com/29359032/228769518-24d10e87-9c45-4e24-b67c-704a38596f5d.png)

It's in the docs :) https://www.npmjs.com/package/lru-cache

jochem-brouwer replied:
Quick note: if you set the cache size to `1` then the tests all fail (this implies that somewhere we use the cache wrong). If you use the current setting, the tests will go out-of-memory.

jochem-brouwer replied:
(at least on the version I tested yesterday)

holgerd77 replied:
Ah, that's really interesting, even 10, 100, and 1000 fail for `npm run tape -- test/execution/vmexecution.spec.ts`, 10000 is the first setting where this goes through.

Will have a look.

holgerd77 replied:
Ah, one thing I am side-realizing while debugging: when we are running the client tests on mainnet, there is always (on every test run) the need to re-fill the state with all these hundres of  genesis/presale accounts, this will very likely slow down test runs significantly. So we might want to switch over to sepolia or so where ever we can.

holgerd77 replied:
Update: ah, so these numbers from above, succeed or not, is likely the number of presale wallets.

holgerd77 replied:
Ok, yeah, it is that the current design is just not working properly if the cache size is lower than number of accounts modified during a block execution.

This is since these modifications are only put into the trie on `flush()` execution, and if there are more puts than the cache can hold in between the old (not yet saved) values are getting out of the cache again before this is persisted.

This would need a substantial cache re-write (again) if this should be tackled, I wouldn't want to do that. But we should definitely point this out in the docs.

holgerd77 replied:
[49df9605133642d40b456505188ea2edad7002ab](https://github.com/ethereumjs/ethereumjs-monorepo/pull/2592/commits/b6eafac9cdd62637a3f7274c924d63d4c75d11d6)

holgerd77 commented:
Ready for review (together with #2618). 🙂 

Do not merge, PRs will be rebased on top of `develop-v7` after review.

acolytec3 reviewed (commented):
Partial review done, will continue on the state manager itself next.

acolytec3 commented on packages/statemanager/src/stateManager.ts:
```diff
@@ -76,8 +121,10 @@ export interface DefaultStateManagerOpts {
 export class DefaultStateManager extends BaseStateManager implements StateManager {
   _trie: Trie
   _storageTries: { [key: string]: Trie }
+  _codeCache: { [key: string]: Buffer }
```
This is remarkable to me that we didn't have a codecache before.  Great add!

acolytec3 commented on packages/statemanager/src/stateManager.ts:
```diff
@@ -87,39 +134,102 @@ export class DefaultStateManager extends BaseStateManager implements StateManage
       const trie = this._trie
-      await trie.del(keyBuf)
+      // This is fixing a bug in the VM GeneralStateTestsRunner passing undefined here for selected accounts
+      // and which breaks when account cache is being deactivated
```
```suggestion
      // and which breaks when account cache is deactivated
```

acolytec3 commented on packages/statemanager/src/stateManager.ts:
```diff
@@ -139,9 +249,15 @@ export class DefaultStateManager extends BaseStateManager implements StateManage
     if (this.DEBUG) {
       this._debug(`Update codeHash (-> ${short(codeHash)}) for account ${address}`)
     }
+    if (!(await this.getAccount(address))) {
```
```suggestion
    if ((await this.getAccount(address) === undefined)) {
```

I'm surprised the linter isn't complaining here about an explicit boolean check.

holgerd77 replied:
Wrong parantheses with await, online code suggestions are always a bit dangerous 😋.

But will generally update, thanks, still not sure about this linter thing.

acolytec3 commented on packages/statemanager/src/stateManager.ts:
```diff
@@ -153,15 +269,25 @@ export class DefaultStateManager extends BaseStateManager implements StateManage
+      return this._codeCache[keyHex]
+    } else {
+      // @ts-expect-error
+      const code = (await this._trie._db.get(key)) ?? Buffer.alloc(0)
```
Just a question about best practice on Typescript errors like this, should we use the above `@ts-expect-error` which makes it not obvious to me what the actual error is (since VS Code respects this and doesn't give you the red squiggles), or should we do `(await (this._trie as any)._db...`?  I'm fine with this approach but maybe you could add a comment explaining what the expected error is (i.e. `db` is a protected member of the `trie` class)

holgerd77 replied:
Just seeing, thanks for updating. 

Update: ah, and just also seeing that we have a public access method for the database in the Trie library (`database()`).

Will just update with this, so can prevent the whole thing.

acolytec3 commented on packages/statemanager/src/stateManager.ts:
```diff
@@ -260,6 +392,9 @@ export class DefaultStateManager extends BaseStateManager implements StateManage
     }
 
     value = unpadBuffer(value)
+    if (!(await this.getAccount(address))) {
```
```suggestion
    if ((await this.getAccount(address)) === undefined) {
```

acolytec3 commented on packages/statemanager/src/stateManager.ts:
```diff
@@ -329,6 +465,9 @@ export class DefaultStateManager extends BaseStateManager implements StateManage
    */
   async getProof(address: Address, storageSlots: Buffer[] = []): Promise<Proof> {
     const account = await this.getAccount(address)
+    if (!account) {
```
Here again, very surprised the linter isn't complaining.  I thought we had a rule for explicit boolean checks where a variable contains either a value or `undefined`?
```suggestion
    if (account === undefined) {
```

acolytec3 commented on packages/statemanager/src/stateManager.ts:
```diff
@@ -465,8 +608,11 @@ export class DefaultStateManager extends BaseStateManager implements StateManage
 
     this._trie.root(stateRoot)
-    this._cache.clear()
+    if (this._cache && clearCache) {
+      this._cache.clear()
```
```suggestion
    if (clearCache) {
      this._cache?.clear()
```

acolytec3 commented on packages/statemanager/src/stateManager.ts:
```diff
@@ -509,17 +655,29 @@ export class DefaultStateManager extends BaseStateManager implements StateManage
-    }
-    if (await this._trie.get(address.buf)) {
-      return true
+    if (this._cache) {
```
```suggestion
    if (this._cache !== undefined) {
```

acolytec3 commented on packages/statemanager/src/cache.ts:
```diff
@@ -1,192 +1,366 @@
   /**
    * Puts account to cache under its address.
-   * @param key - Address of account
+   * @param key - Address of account or undefined if account doesn't exist in the trie
```
Can we rewrite the docs to match the existing function spec?  You now have `address` and `account` where `account` can be either an account or `undefined` so not sure if the above docs still match.

holgerd77 replied:
Generally did some additional round through the code docs.

acolytec3 commented on packages/statemanager/src/cache.ts:
```diff
@@ -1,192 +1,366 @@
   /**
-   * Returns the queried account or an empty account.
+   * Returns the queried account or undefined if account doesn't exist
    * @param key - Address of account
```
```suggestion
   * @param address - Address of account
```

acolytec3 commented on packages/statemanager/src/cache.ts:
```diff
@@ -1,192 +1,366 @@
-   * Returns the queried account or undefined.
-   * @param key - Address of account
+   * Marks address as deleted in cache.
+   * @param key - Address
```
```suggestion
   * @param address - Address
```

acolytec3 commented on packages/statemanager/src/cache.ts:
```diff
@@ -1,192 +1,366 @@
 
   /**
    * Returns true if the key was deleted and thus existed in the cache earlier
    * @param key - trie key to lookup
```
```suggestion
   * @param address - trie key to lookup
```

acolytec3 commented on packages/statemanager/src/cache.ts:
```diff
@@ -1,192 +1,366 @@
   /**
    * Looks up address in cache, if not found, looks it up
    * in the underlying trie.
    * @param key - Address of account
```
```suggestion
   * @param address - Address of account
```

acolytec3 reviewed (commented):
Finished an initial review.  Need to go back over the cache again and really make sure i understand it.

holgerd77 commented on packages/client/lib/rpc/modules/eth.ts:
```diff
@@ -514,6 +513,9 @@ export class Eth {
     await vm.stateManager.setStateRoot(block.header.stateRoot)
     const account = await vm.stateManager.getAccount(address)
+    if (account === undefined) {
+      throw new Error(`could not read account`)
```
TBH, I have no idea why I added this here, that was very much a mistake, yes, will update.

holgerd77 commented on packages/client/test/sync/beaconsync.spec.ts:
```diff
@@ -195,7 +197,7 @@ tape('[BeaconSynchronizer]', async (t) => {
 
   t.test('should extend and set with a valid head', async (t) => {
-    const config = new Config({ transports: [] })
+    const config = new Config({ transports: [], cacheSize: 10000 })
```
Not sure, an eventual more elegant way to solve this need of always passing in the - lowered - cache sizes in the tests might be to default the client to an ORDERED_MAP cache (and so get rid of this size problem) and only from the `bin/cli.ts` instantiation instantiate with LRU for cache_type.

Somewhat of a downside might be that tests are bit less "realistic" when we take another cache type for the tests.

No strong opinion here though, let me know if you have any thoughts or tendencies.

acolytec3 replied:
Not sure if I have an opinion either way.  Doesn't the LRU cache type give significant performance improvements we would when running the client?

holgerd77 replied:
Client should always be LRU - yes - this is just a question if the test setup would be a bit easier (and also somewhat less error prone, if one e.g. forgets to add the lower cache limits for config) if realized in the above-described way.

Anyhow, will leave for now, but still somewhat inclined to do, no super strong opinion though.

holgerd77 commented on packages/evm/src/interpreter.ts:
```diff
@@ -918,6 +928,9 @@ export class Interpreter {
       // update stateRoot on current contract
       const account = await this._eei.getAccount(this._env.address)
+      if (!account) {
+        throw new Error('could not read contract account')
```
Generally - yes - we can use to detect systemic issues early and not have this disguided by account creation (which then goes on with a wrong nonce or so e.g.). That said, I removed most of these patterns (throw) again in the storage cache PR, since these places tend to need some dedicated attention and the initial test failures were too overwhelming.

That often doesn't mean that it would be plain wrong to change, in fact at the end of my storage cache work I re-introduced some of these (also for finding my own bugs in the code).

acolytec3 commented on packages/statemanager/test/cache.spec.ts:
```diff
@@ -2,7 +2,7 @@ import { Trie } from '@ethereumjs/trie'
 import * as tape from 'tape'
 
-import { Cache } from '../src/cache'
+import { Cache, CacheType } from '../src/cache'
```
General question on this, should we be testing with both cache types here (so just run these tests in a loop over eac cache type)?

holgerd77 replied:
Yes, makes sense, have updated.

holgerd77 commented:
Also some procedural note: will keep this "un-rebased" during the review period and then squash the commits from here as Andrew suggested and do a first `master` rebase (primarily to get the snap sync stuff integrated, @scorbajio might need some help there) and then do the `develop-v7` rebase in the next step.

holgerd77 commented:
(also eventually possible: some mixed procedure, so might already start with the rebases during review period and then do the review fixes on top of the rebased and squashed PR. This unsquashed PR should nevertheless remain the main review location)

holgerd77 commented on packages/client/lib/rpc/modules/eth.ts:
```diff
@@ -684,7 +686,10 @@ export class Eth {
-    const account: Account = await vm.stateManager.getAccount(address)
+    const account = await vm.stateManager.getAccount(address)
+    if (account === undefined) {
+      throw new Error(`could not read account`)
```
Same, have updated.

holgerd77 commented on packages/evm/src/evm.ts:
```diff
@@ -615,8 +630,9 @@ export class EVM implements EVMInterface {
         // It is thus an unnecessary default item, which we have to save to dik
         // It does change the state root, but it only wastes storage.
         //await this._state.putContractCode(message.to, result.returnValue)
+        //const account = await this.eei.getAccount(message.to)
```
I can do that but for the other commented out line with `putContractCode()` I am not sure why this is there and if this is maybe a reminder that something is still off here or if this is safe to remove? 🤔

Maybe @jochem-brouwer can shed some light here? Or do you have any context as well?
